build_tools/third_party/swiftshader/build_vk_swiftshader.sh

```diff
@@ -15,51 +15,60 @@
 # limitations under the License.
 set -e
 
-# Call this script **from the project root** to build SwiftShader's Vulkan ICD:
+# Call this script to build SwiftShader's Vulkan ICD:
 #   $ bash build_tools/third_party/swiftshader/build_vk_swiftshader.sh
 #
+# The default installation directory is `${HOME?}/.swiftshader`, but it can be
+# overridden as follows:
+#   $ bash build_tools/third_party/swiftshader/build_vk_swiftshader.sh <swiftshader-install-dir>
+#
 # Note that you will need a working CMake installation for this script to
 # succeed. On Windows, Visual Studio 2019 is recommended.
 #
 # Afterwards, set the VK_ICD_FILENAMES environment variable to the absolute
 # path of the corresponding vk_swiftshader_icd.json manifest file so the Vulkan
-# loader on your system loads it, for example:
-#   Windows
-#   $ set VK_ICD_FILENAMES=C:\dev\iree\build-swiftshader\Windows\vk_swiftshader_icd.json
-#
-#   Linux
-#   $ VK_ICD_FILENAMES=/dev/iree/build-swiftshader/Linux/vk_swiftshader_icd.json
+# loader on your system loads it.
 #
 # See https://vulkan.lunarg.com/doc/view/1.1.70.1/windows/loader_and_layer_interface.html
 # for further details about the Vulkan loader and ICDs.
 
-# Check that we're in the project root so our relative paths work as expected.
-if [[ $(basename "$PWD") != "iree" ]]; then
-    >&2 echo "******************************************************"
-    >&2 echo "* This script should be run from IREE's project root *"
-    >&2 echo "******************************************************"
-    exit 1
-fi
+SWIFTSHADER_COMMIT=6287c18b1d249152563f0cb2d5cb0c6d0eb9e3d6
+SWIFTSHADER_DIR="$(mktemp --directory --tmpdir swiftshader_XXXXXX)"
+SWIFTSHADER_INSTALL_DIR="${1:-${HOME?}/.swiftshader}"
 
-# Generate build system in build-swiftshader/ for third_party/swiftshader/.
-#
+#  Clone swiftshader and checkout the appropriate commit.
+git clone https://github.com/google/swiftshader "${SWIFTSHADER_DIR?}"
+cd "${SWIFTSHADER_DIR?}"
+git pull origin master --ff-only
+git checkout "${SWIFTSHADER_COMMIT?}"
+
+# Install swiftshader in SWIFTSHADER_INSTALL_DIR.
 # Options:
 #   - 64 bit platform and host compiler
 #   - Build Vulkan only, don't build GL
 #   - Don't build samples or tests
-cmake -B build-swiftshader/ \
+
+if [[ -d "${SWIFTSHADER_INSTALL_DIR?}" ]]; then
+  rm -rf "${SWIFTSHADER_INSTALL_DIR?}"
+fi
```

Comment:
This worries me since `SWIFTSHADER_INSTALL_DIR` could be set by a user. If they set it to `~` expecting that to expand to `~/.swiftshader`, this will `rm -rf ~`. How about adding a confirmation first, or always appending `.swiftshader` to the path?

Reply:
I strongly dislike bash lol. I'm not going to write a confirmation loop, so I'll just add an `echo` that you have to delete the directory if you want to write to it.

Reply:
Went for the argument specifies the parent directory route.